File: api/catalog/services/evidence_envelope.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any

from django.db.models import Prefetch

from catalog.models import (
    Asset,
    ClaimEvidence,
    CuratedClaim,
    EvidenceSpan,
)
from catalog.services.semantic_search import viewer_access_statuses
from catalog.services.publication_eligibility import (
    active_catalog_snapshot,
    active_document_revision_q,
    public_editions,
)


@dataclass(frozen=True, slots=True)
class EvidenceEnvelope:
    id: str
    kind: str
    source: dict[str, Any]
    text: str
    locator: dict[str, Any]
    quality: dict[str, Any]
    provenance: dict[str, Any]
    reader_url: str
    pdf_url: str

    def as_dict(self) -> dict[str, Any]:
        return asdict(self)


def evidence_span_envelope(
    span: EvidenceSpan,
    *,
    public_metadata: bool = True,
) -> EvidenceEnvelope:
    revision = span.document_revision
    asset = revision.asset
    edition = asset.edition
    work = edition.work
    page = span.page_number or span.page.index
    snapshot = (
        active_catalog_snapshot(edition, require_fulltext=True)
        if public_metadata
        else {}
    )
    if snapshot:
        authors = [
            str((row.get("person") or {}).get("preferred_name") or row.get("name") or "")
            for row in snapshot.get("contributions") or []
            if isinstance(row, dict)
            and row.get("role") == "author"
            and ((row.get("person") or {}).get("preferred_name") or row.get("name"))
        ]
        work_title = str((snapshot.get("work") or {}).get("title") or "未题名")
    else:
        authors = [
            contribution.person.preferred_name
            for contribution in edition.contributions.all()
            if contribution.approved
            and contribution.role == "author"
            and contribution.person_id
        ]
        work_title = work.title
    focus_id = span.passage_id or span.id
    return EvidenceEnvelope(
        id=str(span.id),
        kind="collection_text",
        source={
            "work_id": str(work.id),
            "work_title": work_title,
            "edition_id": str(edition.id),
            "asset_id": str(asset.id),
            "authors": authors,
            "document_revision_id": str(revision.id),
            "document_revision": revision.revision,
        },
        text=span.original_text,
        locator={
            "page": page,
            "page_id": str(span.page_id),
            "passage_id": str(span.passage_id) if span.passage_id else None,
            "evidence_span_id": str(span.id),
            "printed_page_label": span.printed_page_label,
            "start_offset": span.start_offset,
            "end_offset": span.end_offset,
            "bbox": span.bbox,
            "section": span.section,
        },
        quality={
            "score": span.quality,
            "stale": span.is_stale,
            "stale_reason": span.stale_reason,
        },
        provenance={
            "source_checksum": revision.source_checksum,
            "text_checksum": revision.text_checksum,
            "parser": revision.parser_name,
            "parser_version": revision.parser_version,
            "extraction_method": span.extraction_method or revision.extraction_method,
            "extraction_version": revision.extraction_version,
            "ocr_provider": revision.ocr_provider,
            "ocr_model": revision.ocr_model,
            "ocr_version": revision.ocr_version,
            "ocr": span.ocr_provenance,
        },
        # Reader's passage focus endpoint also accepts EvidenceSpan ids as a
        # locator fallback. This keeps Viewpoint Search and Ask Library on the
        # same verifiable highlight path as full-text search.
        reader_url=f"/reader/{asset.id}?page={page}&passage={focus_id}",
        pdf_url=f"/api/catalog/assets/{asset.id}/manifest/",
    )


def _serialize_curated_claim_groups(
    claims,
    *,
    evidence_attribute: str,
    allowed_kinds: tuple[str, ...],
    public_metadata: bool = True,
) -> dict[str, list[dict[str, Any]]]:
    groups = {kind: [] for kind in allowed_kinds}
    position_map = {
        ClaimEvidence.Role.SUPPORTS: "support",
        ClaimEvidence.Role.OPPOSES: "oppose",
        ClaimEvidence.Role.QUALIFIES: "qualify",
    }
    for claim in claims:
        evidence = []
        positions: list[str] = []
        for link in getattr(claim, evidence_attribute):
            envelope = evidence_span_envelope(
                link.evidence_span,
                public_metadata=public_metadata,
            ).as_dict()
            envelope.update(
                {
                    "claim_role": link.role,
                    "claim_role_label": link.get_role_display(),
                    "claim_confidence": link.confidence,
                }
            )
            evidence.append(envelope)
            position = position_map.get(link.role)
            if position and position not in positions:
                positions.append(position)
        if not evidence:
            continue
        groups[claim.kind].append(
            {
                "id": str(claim.id),
                "kind": claim.kind,
                "kind_label": claim.get_kind_display(),
                "title": claim.title,
                "proposition": claim.proposition,
                "editorial_note": claim.editorial_note,
                "qualifiers": claim.qualifiers,
                "position": positions[0] if len(positions) == 1 else "direct",
                "evidence": evidence,
            }
        )
    return groups


def public_curated_claim_groups(
    *,
    target_field: str,
    target_id,
    allowed_kinds: tuple[str, ...],
) -> dict[str, list[dict[str, Any]]]:
    """Serialize only published curation backed by current public PDF evidence."""

    if target_field not in {"work", "node", "scholar", "topic"}:
        raise ValueError("unsupported CuratedClaim target")
    public_links = (
        ClaimEvidence.objects.filter(
            evidence_span__is_stale=False,
            evidence_span__document_revision__asset__kind=Asset.Kind.NORMALIZED,
            evidence_span__document_revision__asset__status=Asset.Status.READY,
            evidence_span__document_revision__asset__access_status__in=viewer_access_statuses(),
        )
        .filter(
            active_document_revision_q(
                revision_prefix="evidence_span__document_revision"
            )
        )
        .select_related(
            "evidence_span__page",
            "evidence_span__document_revision__asset__edition__work",
        )
        .prefetch_related(
            "evidence_span__document_revision__asset__edition__contributions__person",
        )
        .order_by("sort_order", "created_at")
    )
    claims = (
        CuratedClaim.objects.filter(
            **{
                target_field: target_id,
                "status": CuratedClaim.Status.PUBLISHED,
                "kind__in": allowed_kinds,
                "evidence_links__in": public_links,
            }
        )
        .distinct()
        .prefetch_related(
            Prefetch(
                "evidence_links",
                queryset=public_links,
                to_attr="public_evidence_links",
            )
        )
        .order_by("kind", "sort_order", "created_at")
    )
    if target_field == "work":
        # A newly confirmed claim may already have a canonical published row
        # while the new catalog revision is preparing. Readers keep the
        # previously activated selection until all required consumers finish.
        active_ids = {
            str(identifier)
            for edition in public_editions(require_fulltext=True).filter(work_id=target_id)
            for identifier in (edition.active_catalog_revision.snapshot or {}).get("curated_claim_ids", [])
        }
        claims = claims.filter(pk__in=active_ids)
    return _serialize_curated_claim_groups(
        claims,
        evidence_attribute="public_evidence_links",
        allowed_kinds=allowed_kinds,
    )


def admin_preview_curated_claim_groups(
    *,
    work_id,
    edition_id,
    allowed_kinds: tuple[str, ...],
) -> dict[str, list[dict[str, Any]]]:
    """Include current draft curation for an authenticated page preview."""

    preview_links = (
        ClaimEvidence.objects.filter(
            evidence_span__is_stale=False,
            evidence_span__document_revision__is_active=True,
            evidence_span__document_revision__asset__edition_id=edition_id,
            evidence_span__document_revision__asset__kind=Asset.Kind.NORMALIZED,
            evidence_span__document_revision__asset__status=Asset.Status.READY,
            evidence_span__document_revision__asset__is_current=True,
            evidence_span__document_revision__asset__access_status__in=viewer_access_statuses(
                authenticated=True,
                staff=True,
            ),
        )
        .select_related(
            "evidence_span__page",
            "evidence_span__document_revision__asset__edition__work",
        )
        .prefetch_related(
            "evidence_span__document_revision__asset__edition__contributions__person",
        )
        .order_by("sort_order", "created_at")
    )
    claims = (
        CuratedClaim.objects.filter(
            work_id=work_id,
            status__in=(CuratedClaim.Status.DRAFT, CuratedClaim.Status.PUBLISHED),
            kind__in=allowed_kinds,
            evidence_links__in=preview_links,
        )
        .distinct()
        .prefetch_related(
            Prefetch(
                "evidence_links",
                queryset=preview_links,
                to_attr="preview_evidence_links",
            )
        )
        .order_by("kind", "sort_order", "created_at")
    )
    return _serialize_curated_claim_groups(
        claims,
        evidence_attribute="preview_evidence_links",
        allowed_kinds=allowed_kinds,
        public_metadata=False,
    )

File: web/components/admin/research/evidence-envelope.ts
export type NormalizedEvidenceEnvelope = {
  id: string;
  title: string;
  authors: string[];
  text: string;
  page: string;
  printedPageLabel: string;
  section: string;
  locatorDetails: string[];
  qualityScore: number | null;
  qualityDetails: string[];
  stale: boolean;
  staleReason: string;
  extractionMethod: string;
  extractionVersion: string;
  ocrProvider: string;
  ocrModel: string;
  ocrVersion: string;
  readerUrl: string;
  pdfUrl: string;
  sourceUrl: string;
};

function asRecord(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {};
}

function asString(value: unknown): string {
  if (typeof value === "string") return value.trim();
  if (typeof value === "number") return String(value);
  return "";
}

function firstString(...values: unknown[]): string {
  for (const value of values) {
    const normalized = asString(value);
    if (normalized) return normalized;
  }
  return "";
}

function stringRows(value: unknown): string[] {
  if (Array.isArray(value)) return value.map(asString).filter(Boolean);
  const row = asString(value);
  return row ? [row] : [];
}

function qualityDetails(quality: Record<string, unknown>): string[] {
  const rows: string[] = [];
  const summary = firstString(quality.summary, quality.label);
  if (summary) rows.push(summary);
  if (quality.body_fetched === true) rows.push("已取得正文");
  if (quality.identity_term_present === true) rows.push("身份词已核对");
  const contentType = firstString(quality.content_type);
  if (contentType) rows.push(contentType);
  return [...new Set(rows)];
}

function locatorDetails(locator: Record<string, unknown>): string[] {
  const rows: string[] = [];
  const block = firstString(locator.block_id, locator.text_block_id);
  if (block) rows.push(`区块 ${block}`);
  const start = firstString(locator.start_offset, locator.start);
  const end = firstString(locator.end_offset, locator.end);
  if (start || end) rows.push(`字符 ${start || "?"}–${end || "?"}`);
  if (Array.isArray(locator.bbox) && locator.bbox.length) rows.push(`页内坐标 ${locator.bbox.map(asString).filter(Boolean).join(", ")}`);
  return rows;
}

export function normalizeEvidenceEnvelope(value: unknown, fallbackId = "evidence"): NormalizedEvidenceEnvelope {
  const row = asRecord(value);
  const source = asRecord(row.source);
  const locator = asRecord(row.locator);
  const quality = asRecord(row.quality);
  const provenance = asRecord(row.provenance);
  const nestedOcr = { ...asRecord(row.ocr_provenance), ...asRecord(provenance.ocr) };
  const page = firstString(locator.page, locator.page_number, row.page_number, row.page, row.page_index);
  const asset = firstString(source.asset_id, row.asset_id, row.asset);
  const readerUrl = firstString(row.reader_url, row.reader_href, locator.reader_url) || (asset && page ? `/reader/${encodeURIComponent(asset)}?page=${encodeURIComponent(page)}` : "");
  const qualityValue = quality.score ?? (typeof row.quality === "number" ? row.quality : undefined) ?? row.quality_score ?? row.ocr_quality;
  const parsedQuality = typeof qualityValue === "number" ? qualityValue : Number(qualityValue);
  return {
    id: firstString(row.id, fallbackId),
    title: firstString(source.work_title, source.title, row.source_title, row.work_title, row.title, "馆藏证据"),
    authors: stringRows(source.authors ?? row.authors),
    text: firstString(row.text, row.supporting_text, row.evidence_text, row.text_quote, row.quote, "未提供支撑片段"),
    page,
    printedPageLabel: firstString(locator.printed_page_label, locator.printed_label, row.printed_page_label),
    section: firstString(locator.section, row.section),
    locatorDetails: locatorDetails(locator),
    qualityScore: Number.isFinite(parsedQuality) ? parsedQuality : null,
    qualityDetails: qualityDetails(quality),
    stale: quality.stale === true || row.is_stale === true || row.is_current === false,
    staleReason: firstString(quality.stale_reason, row.stale_reason),
    extractionMethod: firstString(provenance.extraction_method, row.extraction_method),
    extractionVersion: firstString(provenance.extraction_version, row.extraction_version),
    ocrProvider: firstString(provenance.ocr_provider, nestedOcr.provider, row.ocr_provider),
    ocrModel: firstString(provenance.ocr_model, nestedOcr.model, row.ocr_model),
    ocrVersion: firstString(provenance.ocr_version, nestedOcr.version, row.ocr_version),
    readerUrl,
    pdfUrl: firstString(row.pdf_url, row.pdf_href, locator.pdf_url),
    sourceUrl: firstString(row.canonical_url, row.source_url, row.url, source.url, locator.url),
  };
}
